module/executiondatasync/optimistic_sync/pipeline/core.go

```diff
@@ -0,0 +1,305 @@
+			inmemRegisters:                inmemRegisters,
+			inmemEvents:                   inmemEvents,
+			inmemCollections:              inmemCollections,
+			inmemTransactions:             inmemTransactions,
+			inmemResults:                  inmemResults,
+			inmemTxResultErrMsgs:          inmemTxResultErrMsgs,
+		},
+		executionResult: executionResult,
+		header:          header,
+	}
+}
+
+// Download downloads execution data and transaction results error for the block
+//
+// Expected error returns during normal operations:
+// - context.Canceled: if the provided context was canceled before completion
+// - ErrAbandoned: if the core is already abandoned
+func (c *Core) Download(ctx context.Context) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.workingData == nil {
+		return ErrAbandoned
+	}
+
+	c.log.Debug().Msg("downloading execution data")
+
+	g, gCtx := errgroup.WithContext(ctx)
+
+	var executionData *execution_data.BlockExecutionData
+	g.Go(func() error {
+		var err error
+		executionData, err = c.workingData.execDataRequester.RequestExecutionData(gCtx)
+		//  executionData are CRITICAL. Any failure here causes the entire download to fail.
+		if err != nil {
+			return fmt.Errorf("failed to request execution data: %w", err)
+		}
+
+		return nil
+	})
+
+	var txResultErrMsgsData []flow.TransactionResultErrorMessage
+	g.Go(func() error {
+		timeoutCtx, cancel := context.WithTimeout(gCtx, c.workingData.txResultErrMsgsRequestTimeout)
+		defer cancel()
+
+		var err error
+		txResultErrMsgsData, err = c.workingData.txResultErrMsgsRequester.Request(timeoutCtx)
+		if err != nil {
+			// txResultErrMsgsData are OPTIONAL. Timeout error `context.DeadlineExceeded` is handled gracefully by
+			// returning nil, allowing processing to continue with empty error messages data. Other errors still cause
+			// failure.
+			//
+			// This approach ensures that temporary unavailability of transaction result error messages doesn't block
+			// critical execution data processing.
+			if errors.Is(err, context.DeadlineExceeded) {
+				c.log.Debug().
+					Dur("timeout", c.workingData.txResultErrMsgsRequestTimeout).
+					Msg("transaction result error messages request timed out")
+				return nil
+			}
+
+			return fmt.Errorf("failed to request transaction result error messages data: %w", err)
+		}
+		return nil
+	})
+
+	if err := g.Wait(); err != nil {
+		return err
+	}
+
+	c.workingData.executionData = execution_data.NewBlockExecutionDataEntity(c.executionResult.ExecutionDataID, executionData)
+	c.workingData.txResultErrMsgsData = txResultErrMsgsData
+
+	c.log.Debug().Msg("successfully downloaded execution data")
+
+	return nil
+}
+
+// Index retrieves the downloaded execution data and transaction results error messages from the
+// caches and indexes them into in-memory storage.
+//
+// Expected error returns during normal operations:
+// - ErrAbandoned: if the core is already abandoned
+func (c *Core) Index() error {
```

Comment:
I find that this can lock unreasonably long, because we are doing complex database operations. But conceptually, there is no reason to lock: by the time we trigger index, we know the data is there. No more mutations should happen. We should allow other threads to inspect the status of Core.